### DISCUSSION TITLE
inline style syntax issues

### DISCUSSION THREAD

**Discussion by grivo** [plantuml/plantuml#1347] [Announcements]
I can create a relation without declaring participating elements

```
@startuml
:joe: --> (action)
@enduml
```

[link](//www.plantuml.com/plantuml/png/SoWkIImgAStDuRBAoa-rKj3LjLFGI4mkoSpFq-HoICrB0Ke10000)

---

I can set some relation style inline

```
@startuml
:joe: -[#green]-> (action)
@enduml
```
[link](//www.plantuml.com/plantuml/png/SoWkIImgAStDuRBAoa-rKj2DLaulIatDYzMrKz18J2x9pC_Jv798pKi1AW40)

or (this variant seems to me kind of illogical though - and it's better to deprecate it)

```
@startuml
:joe: --> (action) #green
@enduml
```

[link](//www.plantuml.com/plantuml/png/SoWkIImgAStDuRBAoa-rKj3LjLFGI4mkoSpFqrHGJYzAJStZSaZDIm6g0000)

---

I can not set inline style for link and for participating elements other way than declaring elements first alongside with their style, and i can not set stereotype or mark participating element as business valuable other way than declaring this separately as well

code below is best inline solution, imo

```
@startuml
:joe:/ <<xxx>> #green -[#blue]-> (action)/ <<yyy>> #orange
@enduml
```

[link](//www.plantuml.com/plantuml/png/SoWkIImgAStDuRBAoa-rqbUmiQcegB2pKr1EBqfDpLFGZLPEoYbDZTMrKz18J2x9pC_J1AcfhAm4gSalIinBJ-LoICrB0He10000)

desired behaviour

```
@startuml
:joe:/ <<xxx>> #green
(action)/ <<yyy>> #orange
joe -[#blue]-> action
@enduml
```

[link](//www.plantuml.com/plantuml/png/SoWkIImgAStDuRBAoa-rqbUmiQcegB2pKr1EBqfDpUFII4mkoSpFqmH9L5PMWYJoYnBpqbEvWCeLTAELat9AKsDrxHGWwhWSKlDIGAO10000)

---

moreover - i can not distnguish ordinary and business valueble actors / actions in terms of css styling - it may be useful for applying different style onto these elements without a need to apply separate stereotype

**Comment by arnaudroques** [comment]
With [last snapshot](https://github.com/plantuml/plantuml/releases/tag/snapshot), we have fixed:

```
@startuml
<style>
actor {
  BackGroundColor blue;
  business {
    BackGroundColor red;
  }
}
usecase {
  BackGroundColor lightBlue;
  business {
    BackGroundColor orange;
  }
}
</style>
:joe:
:joe2:/
(walk)
(run)/
@enduml
```

and

```
@startuml
:joe:/  --> (action) 
@enduml
```


<hr>

However, we have an minor issue with your proposal:

```
@startuml
:joe:/ <<xxx>> #green -[#blue]-> (action)/ <<yyy>> #orange
@enduml
```

This is because the [following syntax](http://www.plantuml.com/plantuml/umla/SoWkIImgAStDuR8fBgdCIRNZqYjAJL6evkBqoSnBTSxFoIzI00fOSzLosUX3LLXbvQTQAIZewjefQ2GcbsJcvsagsDW0LTdPSJcavgK0JGC0) is actually working:

```
@startuml
<style>
.red {
	LineColor red;
}
</style>
:joe:  --> (action) <<red>>
@enduml
```

So the `<<red>>` stereotype applies here to the link itself, not to `(action)`.

We *may* propose to slightly change the syntax and to move the stereotype on the link **after** the `:`

So to apply stereotype on the link, you would have to use `:joe:  --> (action) : <<red>>`

Your modified example could be:
```
@startuml
:joe:/ <<xxx>> #green -[#blue]-> (action)/ <<yyy>> #orange : some message <<zzz>>
@enduml
```
where `<<xxx>>` applies on `joe`, `<<yyy>>` applies on `action` and `<<zzz>>` applies on the link itself.

This would slightly break existing diagrams that put stereotype on link, though.
What do you think about it?

**Comment by grivo** [comment]
@arnaudroques 

> With [last snapshot](https://github.com/plantuml/plantuml/releases/tag/snapshot), we have fixed:

that's exactly what i want, thanks!

---

> :joe:/ <<xxx>> #green -[#blue]-> (action)/ <<yyy>> #orange : some message <<zzz>>
> 
> where <<xxx>> applies on joe, <<yyy>> applies on action and <<zzz>> applies on the link itself.
> This would slightly break existing diagrams that put stereotype on link, though.
> What do you think about it?

i'd rather say sort of...

as for compatibility - yet we may assume some workaround (hereafter term _styles_ includes possible stereotypes list as well)
 - these variants applly style **on relation only**
     - **:joe:/ -d-> (action)/ <<zzz, uuu, iii>> #blue : some message** - _applicable for a relation that does not have any inline styles but it's own inline style only (as it is for now)_
     - **:joe:/ -d[<<zzz, uuu, iii>> #blue]-> (action)/ : some message** - _applicable for a relation that may have multiple inline styles declared_
 - these variants apply style **on actor**
     - **:joe:/ <<xxx, qqq, ooo>> #green** - _applicable for an actor that does not have any inline styles but it's own inline style only or for an actor declaration apart of relation (as it is for now)_
     - **:joe:/ [<<xxx, qqq, ooo>> #green]** - _applicable for an actor that may have multiple inline styles declared_
 - these variants apply style **on action**
     - **(action)/ <<yyy, hhh, mmm>> #orange** - _applicable for an action that does not have any inline styles but it's own inline style only or for an action declaration apart of relation (as it is for now)_
     - **(action)/ [<<yyy, hhh, mmm>> #orange]** - _applicable for an action that may have multiple inline styles declared_
 - these variants apply style **on action**
 - these examples may apply multiple inline style **on actor, relation and action same time**
     - **:joe:/ [<<xxx, qqq, ooo>> #green] -d[<<zzz, uuu, iii>> #blue]-> (action)/ [<<yyy, hhh, mmm>> #orange] : some message** - _actor, relation and action_
     - **:joe:/ [<<xxx, qqq, ooo>> #green] -d-> (action)/ [<<yyy, hhh, mmm>> #orange] : some message** - _actor and action_
     - **:joe:/ [<<xxx, qqq, ooo>> #green] -d[<<zzz, uuu, iii>> #blue]-> (action)/ : some message** - _actor and relation_
     - **:joe:/ -d[<<zzz, uuu, iii>> #blue]-> (action)/ [<<yyy, hhh, mmm>> #orange] : some message** - _relation and action_

literally - optional square brackets may solve style / stereotype clash issues indeed

**Comment by evantill** [comment]
Maybe we can close this issue and follow the discussion on [gh discussions](https://github.com/plantuml/plantuml/discussions) ?

**Comment by grivo** [comment]
@evantill - do you want me to do this?

what are the chances for this proposal to be implemented then? because current solution with inline syntax seems to be bit incomplete

**Reply by evantill** [reply]
No, you don't have to close it.
I thought it was fixed. Sorry.

Let it open and we will see what we can do.

We are very busy right now, but we will do our best.

**Comment by grivo** [comment]
any news so far?
